test/distributed/test_functional_api.py

```diff
@@ -0,0 +1,249 @@
+    """
+    def test_pg_creation_with_tag(self):
+        my_group, _ = new_subgroups(group_size=2, pg_tag="blu")
+        my_group2, _ = new_subgroups(group_size=2, pg_tag="blu")
+        self.assertEqual(my_group, my_group2)
+
+        my_group3, _ = new_subgroups(group_size=2, pg_tag="blu2")
+        self.assertNotEqual(my_group, my_group3)
+
+        my_group4, _ = new_subgroups(group_size=2)
+        self.assertNotEqual(my_group, my_group4)
+
+        my_group5, _ = new_subgroups(group_size=2)
+        self.assertNotEqual(my_group4, my_group5)
+
+    def test_pg_lookup_roundtrip(self):
+        pg_tag0, _ = new_subgroups(group_size=2, pg_tag="blu")
+        pg_tag1, _ = new_subgroups(group_size=2, pg_tag="blu2")
+        pg_notag0, _ = new_subgroups(group_size=2)
+        pg_notag1, _ = new_subgroups(group_size=2)
+
+        def roundtrip(pg):
+            tag, rankset, _ = ft_c._expand_group(pg)
+            return c10d._try_find_pg_by_ranks_and_tag(tag, rankset)
+
+        self.assertEqual(pg_tag0, roundtrip(pg_tag0))
+        self.assertEqual(pg_tag1, roundtrip(pg_tag1))
+        self.assertEqual(pg_notag0, roundtrip(pg_notag0))
+        self.assertEqual(pg_notag1, roundtrip(pg_notag1))
+
+    def test_pg_lookup_with_tag(self):
+        pg_tag0, _ = new_subgroups(group_size=2, pg_tag="blu")
+        pg_tag1, _ = new_subgroups(group_size=2, pg_tag="bla")
+        pg_notag0, _ = new_subgroups(group_size=2)
+
+        def roundtrip(pg, pg_tag):
+            tag, rankset, _ = ft_c._expand_group(pg, pg_tag)
+            return c10d._try_find_pg_by_ranks_and_tag(tag, rankset)
+
+        self.assertEqual(pg_tag0, roundtrip(pg_tag1, "blu"))
+        self.assertEqual(pg_tag0, roundtrip(pg_notag0, "blu"))
+        # Cannot erase the tag of a PG
+        self.assertEqual(pg_tag0, roundtrip(pg_tag0, ""))
+
+    def test_find_or_create_pg(self):
+        pg = c10d._find_or_create_pg_by_ranks_and_tag("blu", [0, 1, 2, 3], 2)
+        pg_tag0, _ = new_subgroups(group_size=2, pg_tag="blu")
+        self.assertEqual(pg, pg_tag0)
+
+    def test_find_root_pg(self):
+        pg = c10d._try_find_pg_by_ranks_and_tag("", [0, 1, 2, 3])
+        self.assertEqual(dist.group.WORLD, pg)
+
+class TestTraceableCollectives(MultiThreadedTestCase):
+    @property
+    def world_size(self):
+        return 4
+
+    def setUp(self):
+        super().setUp()
+        self._spawn_threads()
+
+    def test_all_reduce_eager(self):
+        tensor = torch.ones([4])
+        mesh = dt.DeviceMesh("cpu", torch.arange(4))
+
+        res = ft_c.all_reduce(tensor, "sum", mesh)
+        self.assertEqual(res, torch.tensor([4, 4, 4, 4], dtype=torch.float))
+
+        mesh = dt.DeviceMesh("cpu", torch.arange(4).view(2, 2))
+        res2 = ft_c.all_reduce(tensor, "sum", (mesh, 1))
+        self.assertEqual(res2, torch.tensor([2, 2, 2, 2], dtype=torch.float))
+
+class TestMetaCollectives(TestCase):
+    def test_all_reduce(self):
+        x = torch.rand((2, 3, 4), device="meta")
```

Comment:
Should we add a test for FakeTensor mode or using Dynamo?

Reply:
dynamo won't work on this PR.  I have a PR to enable it later (with a test).  My PR currently won't perform well but can unblock.  I can optimize if required, but the 'real' fix is to wait for Brian's change to support tensor subclasses

https://github.com/pytorch/pytorch/pull/94440